marshal.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+)
+
+// Marshal takes the input structure and creates an http request.
```

Comment:
//  Marshal is the counterpart of Unmarshal. It takes information from
	//  x, which must be a pointer to a struct, and returns an HTTP request
	//  using the given method that holds all of the information
	// 
	//  The HTTP request will use the given method.  Named fields in the given
	//  baseURL will be filled out from "path"-tagged fields in x to form the
	//  URL path in the returned request.  These are specified as for httprouter.
	// 
	//  If a field is of type string or []string, the value of the field will
	//  be used directly; otherwise if implements encoding.TextMarshaler, that
	//  will be used to marshal the field, otherwise fmt.Sprint will be used.
	// 
	//  For example, this code:
	//  
	//  	type UserDetails struct {
	//  		Age int
	//  	}
	// 	type Test struct {
	// 		Username string `httprequest:"user,path"`
	// 		ContextId int64 `httprequest:"context,form"`
	// 		Details UserDetails	`httprequest:",body"`
	//  	}
	// 	req, err := Marshal("GET", "http://example.com/users/:user/details", &Test{
	// 		Username: "bob",
	// 		ContextId: 1234,
	// 		Details: UserDetails{
	// 			Age: 36,
	// 		}
	// 	})
	// 	if err != nil {
	// 		...
	// 	}
	// 
	// will produce an HTTP request req with a URL of
	// http://example.com/users/bob/details?context=1234 and a JSON-encoded
	// body holding `{"Age":36}`.
	// 
	// It is an error if there is a field specified in the URL that is not
	// found in x.

?